### PR TITLE
Compactor: Adding minimum retention flag validation for downsampling retention

### PR DESCRIPTION
**Pull request by moadz**
* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes
* [x] Added 40h minimum raw resolution retention if downsampling enabled and `--retention.resolution-raw` > 0 
* [x] Added 10d minimum 5m resolution retention if downsampling enabled and `--retention.resolution-5m` > 0
* [x] Added 10d minimum 1h resolution retention if downsampling enabled and `--retention.resolution-1h` > 0
* [x] Added clarifying docs around why downsampling time delay  

## Verification
* Ran locally with e2e compactor tests with flags enabled below set retention to test flag validation 

Changes based on: https://github.com/thanos-io/thanos/issues/813 

tl;dr: For downsampling to happen, blocks must be pass a certain age for the preceding resolution. This was always the case in [Compactor](https://github.com/thanos-io/thanos/blob/c31cb08f1b0cf9ad7f23a57210812499a8fa0764/cmd/thanos/downsample.go#L265-L318) but not transparently to users. As such, it's very easy to set up retention in such a way that disrupts downsampling or leads to accidental data loss.

### CODE REVIEW

File: docs/components/compact.md [hunk 1]
```diff
@@ -2,10 +2,9 @@
 
 The `thanos compact` command applies the compaction procedure of the Prometheus 2.0 storage engine to block data stored in object storage. It is generally not semantically concurrency safe and must be deployed as a singleton against a bucket.
 
-Compactor is also responsible for downsampling of data:
-
-* Creating 5m downsampling for blocks larger than **40 hours** (2d, 2w)
-* Creating 1h downsampling for blocks larger than **10 days** (2w)
+Compactor is also responsible for downsampling of data. There is a time delay before downsampling at a given resolution is possible. This is necessary because downsampled chunks will have fewer samples in them, and as chunks are fixed size, data spanning more time will be required to fill them.
```
Comment (moadz) [inline docs/components/compact.md:5]: I'm inferring this from the code, 2hr 15s sampling to 20hrs 5min sampling (~factor of 10). It mentions that we want to produce at least 2 chunks for each downsampling pass, unsure of any other reason for why this would be the case.

File: cmd/thanos/compact.go [hunk 2]
```diff
@@ -388,12 +388,24 @@ func runCompact(
 	}
 
 	if retentionByResolution[compact.ResolutionLevelRaw].Seconds() != 0 {
+		// If downsampling is enabled, error if raw retention is not sufficient for downsampling to occur (upper bound 10 days for 1h resolution)
+		if !conf.disableDownsampling && retentionByResolution[compact.ResolutionLevelRaw].Seconds() < downsample.ResLevel1DownsampleRange {
+			return errors.New("raw resolution must be higher than the minimum block size after which 1h resolution downsampling will occur (40 hours)")
```
Comment (yeya24) [inline cmd/thanos/compact.go:393]: I don't get why raw block retention is related to 1h downsampled block. I think 1h downsampled block is created from 5m downsampled block. Can you please explain it?

File: cmd/thanos/compact.go [hunk 3]
```diff
@@ -388,12 +388,24 @@ func runCompact(
...
+		}
 		level.Info(logger).Log("msg", "retention policy of raw samples is enabled", "duration", retentionByResolution[compact.ResolutionLevelRaw])
 	}
 	if retentionByResolution[compact.ResolutionLevel5m].Seconds() != 0 {
+		// If retention is lower than minimum downsample range, then no downsampling at this resolution will be persisted
+		if !conf.disableDownsampling && retentionByResolution[compact.ResolutionLevel5m].Seconds() < downsample.ResLevel2DownsampleRange {
+			return errors.New("5m resolution retention must be higher than the minimum block size after which 5m resolution downsampling will occur (10 days)")
```
Comment (yeya24) [inline cmd/thanos/compact.go:400]: 5m => 1h?

File: cmd/thanos/compact.go [hunk 4]
```diff
@@ -388,12 +388,24 @@ func runCompact(
...
+		}
 		level.Info(logger).Log("msg", "retention policy of 5 min aggregated samples is enabled", "duration", retentionByResolution[compact.ResolutionLevel5m])
 	}
 	if retentionByResolution[compact.ResolutionLevel1h].Seconds() != 0 {
+		// If retention is lower than minimum downsample range, then no downsampling at this resolution will be persisted
+		if !conf.disableDownsampling && retentionByResolution[compact.ResolutionLevel1h].Seconds() < downsample.ResLevel2DownsampleRange {
```
Comment (yeya24) [inline cmd/thanos/compact.go:406]: I think this check is not needed.

File: pkg/compact/downsample/downsample.go [hunk 5]
```diff
@@ -35,8 +35,8 @@ const (
 
 // Downsampling ranges i.e. minimum block size after which we start to downsample blocks (in seconds).
 const (
-	DownsampleRange0 = 40 * 60 * 60 * 1000      // 40 hours.
-	DownsampleRange1 = 10 * 24 * 60 * 60 * 1000 // 10 days.
+	ResLevel1DownsampleRange = 40 * 60 * 60 * 1000      // 40 hours.
+	ResLevel2DownsampleRange = 10 * 24 * 60 * 60 * 1000 // 10 days.
```
Comment (yeya24) [inline pkg/compact/downsample/downsample.go:39]: 👍

File: test/e2e/e2ethanos/services.go [hunk 6]
```diff
@@ -735,7 +735,6 @@ func NewCompactor(e e2e.Environment, name string, bucketConfig client.BucketConf
 				"--selector.relabel-config": string(relabelConfigBytes),
 				"--wait":                    "",
 			}), extArgs...)...),
-			Readiness:        e2e.NewHTTPReadinessProbe("http", "/-/ready", 200, 200),
```
Comment (yeya24) [inline test/e2e/e2ethanos/services.go:738]: Is this change related to this pr?

File: cmd/thanos/compact.go (hunk 2 above)
Comment (moadz) [reply cmd/thanos/compact.go:393]: Copy pasta typo, fixing.

File: cmd/thanos/compact.go (hunk 3 above)
Comment (moadz) [reply cmd/thanos/compact.go:400]: Fixed.

File: cmd/thanos/compact.go (hunk 4 above)
Comment (moadz) [reply cmd/thanos/compact.go:406]: My rationale is if 1h downsampling only happens after 10 days, setting the retention to <10 days means 1h downsampling will never happen. 

Happy to remove, but IMHO it helps to avoid misusing downsampling as a feature.

File: test/e2e/e2ethanos/services.go (hunk 6 above)
Comment (moadz) [reply test/e2e/e2ethanos/services.go:738]: Nope, was just testing the change with this, slipped into the PR. Reverted.

File: cmd/thanos/compact.go (hunk 4 above)
Comment (moadz) [reply cmd/thanos/compact.go:406]: Removed.

**Comment by moadz** [comment]
> I think is going to be a really nice improvement 👏 However I'm still seeing some extra commits in the PR, shall we remove those? Happy to re-review afterwards again!

@matej-g just trying to fix some flakey docs tests by rebasing on master. All current feedback has been addressed pending re-review by @yeya24

File: docs/components/compact.md (hunk 1 above)
Comment (bwplotka) [reply docs/components/compact.md:5]: 👍🏽

**Review by bwplotka — APPROVED** [review]
Thanks, LGTM!
